File: src/leetcode/h76/MinimumWindowSubstring.js
/**
 * LeetCode 76
 *
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function (s, t) {
    let map = new Map;
    for (let i = 0; i < t.length; i++) {
        if (map.has(t[i])) {
            map.set(t[i], map.get(t[i]) + 1);
        } else {
            map.set(t[i], 1);
        }
    }

    let remaining = t.length;
    let minLen = Number.MAX_SAFE_INTEGER;
    let minRight;
    let left = 0;
    let right = 0;
    while (right < s.length) {
        let rightChar = s.charAt(right);
        if (map.has(rightChar)) {
            let count = map.get(rightChar);
            if (count > 0) {
                remaining--;
            }
            map.set(rightChar, --count);
        }
        right++;

        while (remaining === 0) {
            let leftChar = s.charAt(left);
            if (map.has(leftChar)) {
                let count = map.get(leftChar);
                if (count === 0) {
                    let newLen = right - left;
                    if (newLen < minLen) {
                        minLen = newLen;
                        minRight = right;
                    }
                    remaining++;
                }
                map.set(leftChar, ++count);
            }
            left++;
        }
    }

    if (minLen === Number.MAX_SAFE_INTEGER) {
        return "";
    }

    let startIndex = minRight - minLen;
    return s.substring(startIndex, minRight);
};

let s = "ADOBECODEBANC";
let t = "ABC";
console.log(minWindow(s, t));

s = "a";
t = "aa";
console.log(minWindow(s, t));

File: src/leetcode/h76/MinimumWindowSubstring.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package leetcode.h76;

import java.util.HashMap;
import java.util.Map;

/**
 * LeetCode 76
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/08/01 15:18
 */
public class MinimumWindowSubstring {
    public String minWindow(String s, String t) {
        Map<Character, Integer> charMap = new HashMap<>();
        for (int i = 0; i < t.length(); i++) {
            int cnt = charMap.getOrDefault(t.charAt(i), 0);
            charMap.put(t.charAt(i), cnt + 1);
        }

        int slow = 0;
        int fast = 0;
        int minLeft = slow;
        int minLen = Integer.MAX_VALUE;
        int remaining = t.length();
        while (fast < s.length()) {
            char char1 = s.charAt(fast);
            if (charMap.containsKey(char1)) {
                int fastCharCnt = charMap.get(char1);
                if (fastCharCnt > 0) {
                    remaining--;
                }
                charMap.put(char1, fastCharCnt - 1);
            }
            fast++;

            while (remaining == 0) {
                if (fast - slow < minLen) {
                    minLen = fast - slow;
                    minLeft = slow;
                }
                char char2 = s.charAt(slow);
                if (charMap.containsKey(char2)) {
                    int slowCharCnt = charMap.get(char2);
                    if (slowCharCnt >= 0) {
                        remaining++;
                    }
                    charMap.put(char2, slowCharCnt + 1);
                }
                slow++;
            }
        }
        return minLen == Integer.MAX_VALUE ? "" : s.substring(minLeft, minLeft + minLen);
    }
}
